src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp

```diff
@@ -166,7 +179,8 @@ void ExecuteScalarSubqueriesMatcher::visit(const ASTSubquery & subquery, ASTPtr
         auto lit = std::make_unique<ASTLiteral>((*scalar.safeGetByPosition(0).column)[0]);
         lit->alias = subquery.alias;
         lit->prefer_alias_to_column_name = subquery.prefer_alias_to_column_name;
-        ast = addTypeConversionToAST(std::move(lit), scalar.safeGetByPosition(0).type->getName());
+        /// If only analyze was requested the expression is not suitable for constant folding, disable it.
+        ast = addTypeConversionToAST(std::move(lit), scalar.safeGetByPosition(0).type->getName(), suitable_for_const_folding);
```

Comment:
By the way, why do we even need a special branch that runs only for `EXPLAIN SYNTAX`? This means that `EXPLAIN SYNTAX` doesn't even show the query that is going to run, but some special query that doesn't exist in any other case. It doesn't make sense. What breaks if we remove `data.only_analyze` in the above `if`, and take the other branch that replaces the query with `__getScalar`?

Reply:
I mean, maybe the natural volatile wrapper is the `__getScalar` itself?

Reply:
>By the way, why do we even need a special branch that runs only for EXPLAIN SYNTAX?

This PR #10603 has good explanation.

Reply:
Can't say I really understood what it does...

But OK, another thing -- do we *have* to perform constant folding for "real" constants in this case? Maybe we can disable it unconditionally for the `data.only_analyze` branch? This way your patch becomes just one `makeASTFunction` call.

It probably shouldn't be in `addTypeConversionToAST`, it is a generic function.

Reply:
>Can't say I really understood what it does...

IIUC then the problem is only with `CREATE MATERIALIZED VIEW ... AS` query, i.e. `CREATE MATERIALIZED VIEW dst_mv Engine=Memory as select *, (SELECT count() FROM src_table) AS cnt FROM src_table` (here it should not materialize `cnt`)

>But OK, another thing -- do we have to perform constant folding for "real" constants in this case? Maybe we can disable it unconditionally for the data.only_analyze branch? This way your patch becomes just one makeASTFunction call.

This will simplify the code a lot, the initial reason for that complexity to make `EXPLAIN SYNTAX` looks more similar to what `SELECT` will do

And for the following case it will not eliminate redundant `WHERE subquery = 1` as before:

```sql
EXPLAIN SYNTAX SELECT * FROM (SELECT 1 AS id, (SELECT 1) as subquery) WHERE subquery = 1
```

>It probably shouldn't be in addTypeConversionToAST, it is a generic function.

Yes, now (no need to change the function name) it can be moved out of that helper.

Reply:
@akuzm the question is do we want to preserve that case or not. If not I have simpler patch

Reply:
Let's go with the simpler one. Now we are in a weird situation where `explain syntax` sometimes matches the real query, and sometimes we have a special branch for it, that doesn't match anything else, so the only thing it explains is itself :)

Reply:
As I understand, with the simpler patch it will always have a discrepancy w/the real query, but at least in a more predictable way.

Reply:
>Let's go with the simpler one

OK

>Now we are in a weird situation where explain syntax sometimes matches the real query, and sometimes we have a special branch for it, that doesn't match anything else, so the only thing it explains is itself :)

But only sometimes, so I wouldn't count him out yet :)

>As I understand, with the simpler patch it will always have a discrepancy w/the real query, but at least in a more predictable way.

For constants - yes